core/src/main/java/org/apache/gravitino/storage/relational/service/UserMetaService.java

```diff
@@ -221,6 +222,22 @@ public <E extends Entity & HasIdentifier> UserEntity updateUser(
     return newEntity;
   }
 
+  public List<UserEntity> listUsersByNamespace(Namespace namespace) {
+    AuthorizationUtils.checkUserNamespace(namespace);
+
+    List<UserEntity> userEntities = Lists.newArrayList();
+    String metalakeName = namespace.level(0);
+    List<UserPO> userPOs =
+        SessionUtils.getWithoutCommit(
+            UserMetaMapper.class, mapper -> mapper.listUserPOsByMetalake(metalakeName));
+
+    for (UserPO userPO : userPOs) {
+      List<RolePO> rolePOs = RoleMetaService.getInstance().listRolesByUserId(userPO.getUserId());
+      userEntities.add(POConverters.fromUserPO(userPO, rolePOs, namespace));
+    }
```

Comment:
Do you really want to return all the roles for all the users? This will be extremely cost, is that really the frontend needs?

Reply:
Now, we adopts the lazily loading information about users/roles.